### PR TITLE
Bump euclid to 0.14.x.

### PR DESCRIPTION
**Pull request by nical**
- [x] `./mach build -d` does not report any errors (kinda, need webrender published and Cargo.toml fixed up)
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [ ] There are tests for these changes OR
- [x] These changes do not require tests because it is a refactoring in which the difference is mostly a compile-time/strong-typing thing with no change to the logic.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17184)

### CODE REVIEW

**Comment by nical** [comment]
This requires webrender the equivalent WebRender and rust-azure changes to be merged/published, in order to build (I did get it to work locally), but the review is going to be a lot of work so I am submitting this early for people to start looking at it. This pull request is enormous, and I am truly sorry for this.
We made some major changes to euclid, in particular the separation of points and vector types. See https://github.com/servo/euclid/pull/159.

cc @nox

**Comment by asajeffrey** [comment]
Gosh, you're not kidding about this being a big change!

Are most of the changes to euclid naming? Would it be possible/useful to publish an intermediate version of euclid with both naming schemes?

**Comment by nical** [comment]
The biggest change is the separation between vector and points. The rest of the diff is trivial changes like renaming Matrix2D/4D into Transform2D/3D and cleaning up the namespaces (`use euclid::rect::Rect;` becomes `use euclid::Rect`).
There is also two commits about SVG and animations that are not actually from this PR but that appear in the diff. After I have rebased this, they will disappear from the diff.

Once I have rebased this patch though, most of the non-trivial changes won't be in the diff, and if you want I can roll back the the Matrix/Transform renaming (this specific part doesn't occupy much room in the diff though) since I left an alias for the old name in euclid marked as deprecated, so it can be postponed if servo build does not deny warnings.

For the other changes, I am sorry but I spent way too much time getting this euclid bump into all of the servo crates outside the main repo, getting reviews an getting things published on crates in the correct order. I ended up doing mostly that for more than a week and I won't do it again. Had I known how much work this would be, I would not have made these changes to euclid.

**Comment by nical** [comment]
deprecation warnings). The total diff is a lot less scary to look at now.
Let me know if you prefer things squashed as is and move the matrix/transform stuff to a separate PR or if you want it in this one.

I kept another commit separated: the one that turns `BaseFlow::stacking_relative_position` into a vector instead of a point. For this one I wasn't 100% sure whether it made most sense as a vector or a point (short of knowing the layout code well enough), so it is still in its own commit which can be removed easily if need be.

File: Cargo.toml [hunk 1]
```diff
@@ -19,3 +19,7 @@ opt-level = 3
 # crate, add that here. Use the form:
 #
 #     "<crate>:<version>" = { path = "/path/to/local/checkout" }
+
+"https://github.com/servo/webrender#0.40.0" = { path = "../webrender/webrender/" }
+"https://github.com/servo/webrender#webrender_traits:0.40.0" = { path = "../webrender/webrender_traits/" }
+"https://github.com/servo/rust-azure#azure:0.16.0" = { path = "../rust-azure/" }
```
Comment (mrobinson) [inline Cargo.toml]: Now that azure has the WebRender bump as well, I guess this can be removed.

File: components/layout/fragment.rs [hunk 2]
```diff
@@ -2861,11 +2861,11 @@ impl Fragment {
         let transform_origin_z = transform_origin.depth.to_f32_px();
 
         let pre_transform = Matrix4D::create_translation(transform_origin_x,
-                                                         transform_origin_y,
-                                                         transform_origin_z);
+                                                            transform_origin_y,
+                                                            transform_origin_z);
```
Comment (asajeffrey) [inline components/layout/fragment.rs]: Nit: I think our indentation style is the old version of the code.

File: components/layout/fragment.rs [hunk 3]
```diff
@@ -3099,9 +3099,9 @@ impl Overflow {
         self.paint = self.paint.union(&other.paint);
     }
 
-    pub fn translate(&mut self, point: &Point2D<Au>) {
-        self.scroll = self.scroll.translate(point);
-        self.paint = self.paint.translate(point);
+    pub fn translate(&mut self, by: &Vector2D<Au>) {
```
Comment (asajeffrey) [inline components/layout/fragment.rs]: This makes a lot more sense with the new type name!

File: components/layout/webrender_helpers.rs [hunk 4]
```diff
@@ -86,12 +86,22 @@ trait ToPointF {
     fn to_pointf(&self) -> webrender_traits::LayoutPoint;
 }
 
+trait ToVectorF {
+    fn to_vectorf(&self) -> webrender_traits::LayoutVector2D;
```
Comment (asajeffrey) [inline components/layout/webrender_helpers.rs]: Nit: the function name is not very self-explanatory.

**Review by asajeffrey — COMMENTED** [review]
Looks perfectly sensible, and a lot of work! Just a couple of nits.

**Comment by SimonSapin** [comment]
Looks good to me.

---

Reviewed 39 of 87 files at r1, 41 of 51 files at r4, 9 of 9 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[components/canvas/canvas_paint_thread.rs, line 651 at r5](https://reviewable.io:443/reviews/servo/servo/17184#-KmWR55uRM35K1KZnz6X:-KmWR55uRM35K1KZnz6Y:bkww5cy) ([raw file](https://github.com/servo/servo/blob/cee0715de90f86cae0ff98f73a6724bfbf045abf/components/canvas/canvas_paint_thread.rs#L651)):*
> ```Rust
> 
>         // Step 6.
>         let dest_rect = dirty_rect.translate(&offset.to_vector()).to_i32();
> ```

Should `offset` be a vector to start with?

---

*[components/script/dom/document.rs, line 869 at r5](https://reviewable.io:443/reviews/servo/servo/17184#-KmWQwUTl8NwFar37oeQ:-KmWQwUTl8NwFar37oeR:b-cxdltz) ([raw file](https://github.com/servo/servo/blob/cee0715de90f86cae0ff98f73a6724bfbf045abf/components/script/dom/document.rs#L869)):*
> ```Rust
>                 let rect = iframe.upcast::<Element>().GetBoundingClientRect();
>                 let child_origin = Point2D::new(rect.X() as f32, rect.Y() as f32);
>                 let child_point = client_point - child_origin.to_vector();
> ```

Could `child_origin` be a vector to start with? (Here and 3 others in this file.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/17184#-:-KmWSpCNVaojlpJzmTuG:b9j9n43)*

**Comment by nical** [comment]
The review comments have been addressed, should we land this now?

**Comment by SimonSapin** [comment]
Thank you!

@bors-servo r+

---

Reviewed 3 of 13 files at r6, 9 of 9 files at r7, 1 of 10 files at r8, 9 of 9 files at r9.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/17184#-:-KmakHE8gEg37y6ZxOHm:b-cr0ec)*

**Comment by SimonSapin** [comment]
@bors-servo r+

---

Reviewed 15 of 24 files at r10, 9 of 9 files at r11.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/17184#-:-Kmay8YlSCsgRvndNCsi:bkg9nno)*

**Pull request merged**